plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/performance/MapOrSetKeyInspection.java

```diff
@@ -0,0 +1,191 @@
+      final PsiTypeElement typeElement = variable.getTypeElement();
+      if (typeElement == null) {
+        return;
+      }
+      final PsiType type = typeElement.getType();
+      if (!(type instanceof PsiClassType)) {
+        return;
+      }
+      final PsiJavaCodeReferenceElement referenceElement =
+        typeElement.getInnermostComponentReferenceElement();
+      if (referenceElement == null) {
+        return;
+      }
+      final PsiClassType classType = (PsiClassType)type;
+      final PsiClass aClass = classType.resolve();
+
+      final ClassType collectionType = getClassType(aClass);
+      if (collectionType == ClassType.OTHER) {
+        return;
+      }
+
+      final PsiReferenceParameterList parameterList =
+        referenceElement.getParameterList();
+      if (parameterList == null ||
+          parameterList.getTypeParameterElements().length == 0) {
+        final PsiMember member =
+          PsiTreeUtil.getParentOfType(variable, PsiMember.class);
+        if (member == null) {
+          return;
+        }
+        final ValueAddedVisitor visitor =
+          new ValueAddedVisitor(variable, collectionType);
+        member.accept(visitor);
+        if (visitor.isValueOfInterestAdded()) {
+          registerVariableError(variable, collectionType);
+        }
+        return;
+      }
+      final PsiType[] typeArguments = parameterList.getTypeArguments();
+      boolean triggerInspection = false;
+
+      if (typeArguments.length < 1 || typeArguments[0] == null) {
+        return;
+      }
+
+      PsiType typeArgument = typeArguments[0];
+
+      if (shouldTriggerOnKeyType(typeArgument)) {
+        triggerInspection = true;
+      }
+
+      if (!triggerInspection) {
+        return;
+      }
+      registerVariableError(variable, collectionType);
+    }
+  }
+
+  private class ValueAddedVisitor
+    extends JavaRecursiveElementWalkingVisitor {
+
+    private final PsiVariable variable;
+    private final ClassType collectionType;
+    private boolean valueOfInterestAdded;
+
+    ValueAddedVisitor(PsiVariable variable,
+                      ClassType collectionType) {
+      this.variable = variable;
+      this.collectionType = collectionType;
+    }
+
+    @Override
+    public void visitMethodCallExpression(
+      PsiMethodCallExpression expression) {
+      if (valueOfInterestAdded) {
+        return;
+      }
+      super.visitMethodCallExpression(expression);
+      final PsiReferenceExpression methodExpression =
+        expression.getMethodExpression();
+      final PsiExpression qualifierExpression =
+        methodExpression.getQualifierExpression();
+      if (!(qualifierExpression instanceof PsiReferenceExpression)) {
+        return;
+      }
+      final PsiReferenceExpression referenceExpression =
+        (PsiReferenceExpression)qualifierExpression;
+      @NonNls final String methodName =
+        methodExpression.getReferenceName();
+      if (collectionType == ClassType.SET &&
```

Comment:
addAll with a collection of interested type should trigger a warning as well, probably...